Replace --collectors.enabled with per-collector flags

carlpett opened the pull request:
@SuperQ mentioned that after the switch to kingpin, now would be a good time to also replace `--collectors.enabled` with individual flags for each collector, since there is a breaking change anyway.

This PR does that. The flags are now `--collector.<name>.enabled` (or `--no-collector.<name>.enabled` to disable a default collector), but if there are shorter alternatives that might be good to consider.

This PR also moves the `NodeExporter` type into the collector package, as suggested by @SuperQ to follow [mysqld_exporter #209](https://github.com/prometheus/mysqld_exporter/pull/209).

It is possible to remove a few more exported symbols from the `collector` package, but I'd like some opinions on if that would be positive or not. Making `Factories` internal seems like a reasonable thing, for instance, but then we'd need to make `--collectors.print` call into some new function.

carlpett commented:
Working on the bash-fu to fix the e2e test script...

SuperQ commented on end-to-end-test.sh:
```diff
@@ -76,7 +83,8 @@ fi
   --collector.procfs="collector/fixtures/proc" \
   --collector.sysfs="collector/fixtures/sys" \
-  --collectors.enabled="$(echo ${collectors} | tr ' ' ',')" \
+  $(for c in $enabled_collectors; do echo --collector.$c.enabled  ; done) \
```
Minor bash style thing here, for consistency I prefer to use `${var_name}`.

carlpett replied:
Sure, fixed!

SuperQ commented:
This is great, thanks for taking care of this.

discordianfish commented:
Overall looking good, thanks!
One concern though: If a user wants a specific set of collectors enable, they would need to keep track of the currently enabled defaults and disabled the ones they don't need selectively, right?
I think we should have a way to only explicitly enable a list of collectors and I don't think this is possible with this right now. Maybe we could have a flag to disable the defaults? E.g `--no-collector.defaults.enabled` would disable all default-enabled collectors so you can specify explicitly which ones you want to use.

SuperQ commented:
@discordianfish Either way, the user has to track collectors between releases.  I'd rather keep this simple if possible.

I suggest we file an issue for "disable everything flag" and see if anyone really needs it.

grobie commented:
I have similar concerns to @discordianfish and I'm not sure the current change is a clear win in usability. Given PromCon and related preparations, I won't have time to research for ideas in similar setups and comment before next week.

SuperQ commented:
For what it's worth, this is the same style as we use on the `mysqld_exporter`, and I don't recall any requests for "disable everything by default".

To, me, there are some clear usability wins.  The reasoning id detailed in https://github.com/prometheus/node_exporter/issues/216.

discordianfish commented:
@SuperQ But as right now, without this PR you can explicitly set the collectors you want to use. With this change this isn't possible anymore which is why I believe we need to change this before merging this PR. Otherwise I would agree that such improvements could be made later.

The mysql-exporter is IMO a different story where the need to explicitly enable a specific set of option isn't as important.

carlpett commented:
>One concern though: If a user wants a specific set of collectors enable, they would need to keep track of the currently enabled defaults and disabled the ones they don't need selectively, right?

@discordianfish Yes, pretty much. Which is the inverse of the current issue - if I just want to run with the defaults and add a few more, I will have to pull out the defaults, add the optional ones I need, and then keep that list up to date as I update the exporter. 

Empirically, this need is more common requirement (the defaults are fairly uncontroversial, but I want to add one or two of the non-defaults), but maybe the groups I've worked with are outliers in this? 

Anyway, I don't think it is unreasonable to have a "disable all" flag which then lets the user enable the specific ones with `--collector.<name>.enabled` for those cases where someone needs to be very precise about what collectors are enabled (although if one of the default collectors are troublesome in some cases, should it really be default?).

grobie commented:
> although if one of the default collectors are troublesome in some cases, should it really be default?

They shouldn't if they really cause trouble. The most common use case to disable default collectors is to reduce the amount of metrics produced in space limited setups.

matthiasr commented:
+1 for having a way to enable only specific collectors. We enable almost
all new collectors by default these days. Multiple people have spoken up
here with a need to contain this, so I don't see the need for a separate
issue.

On Tue, Aug 15, 2017, 09:58 Tobias Schmidt <notifications@github.com> wrote:

> although if one of the default collectors are troublesome in some cases,
> should it really be default?
>
> They shouldn't if they really cause trouble. The most common use case to
> disable default collectors is to reduce the amount of metrics produced in
> space limited setups.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/node_exporter/pull/640#issuecomment-322402812>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAICBrPq8nr3b4BpQzWiX-AwJbmemAm8ks5sYU-LgaJpZM4O1qSn>
> .
>

discordianfish commented:
Yes, also thinking about limit-number-of-metrics usecase. So we agree about a flag like `--no-collector.defaults.enabled`? Or other preferences?

SuperQ commented:
How about `--disable-collector-defaults`, with the default of `false`.

matthiasr commented:
I like this one better. no-collector.defaults.enabled is really hard to
parse.

On Tue, Aug 15, 2017 at 9:28 AM Ben Kochie <notifications@github.com> wrote:

> How about --disable-collector-defaults, with the default of false.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/node_exporter/pull/640#issuecomment-322422806>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAICBvDgteHFklgvxIdeWORfzOfj9Rs7ks5sYWTHgaJpZM4O1qSn>
> .
>

discordianfish commented:
I liked the consistency but sure that name work

SuperQ commented:
@carlpett Would you add a flag `--disable-collector-defaults` to this PR that sets all collectors to false unless explicitly enabled?

carlpett commented:
@superq Sorry for not responding! I've been travelling without internet access since submitting the PR, but I'll fix that flag as soon as I have access to laptop and wifi again :)

SuperQ commented:
No worries, I was just wanting to ping. 😃

carlpett commented:
Hm, so actually this isn't as easy as I would have first thought. Currently the "enabledness" of collectors is tracked in a `map[string]*bool`, pointed to by the flag, and the default state is set while setting up the flag. This means that changing the default state via a flag is hard - the setting of default state is done before parsing flags.
One potential solution would be to add another map for default state and then copy that over to the enabled-map if the defaults are not disabled. Any better suggestions?

discordianfish commented:
I realized that too when looking at this. Don't really have a better answer. Maybe we should check out how others are doing this.

carlpett commented:
Finally got my laptop back... I split the bool into two, so it is possible to track if the user has chosen to enable it and the default state separately.
I named the flag `--collectors.disable-defaults` to keep it more in line with the general namespacing. What do you think? If you prefer `--disable-collector-defaults` I'll change it to that.

carlpett commented:
Huh, and apparently I broke some of the previous functionality when implementing it. Whoops. I even added tests, but forgot this case.
Fixing it.

carlpett commented:
Have been a bit short on time, so took a bit longer than planned, but it should work now. The flag state management does get a bit verbose, but we need to somehow track all of these things:
1. The default state of the collector
2. If the flag has been given on the commandline
3. If so, was it given as enable or disable?

I hope the flow is understandable from a quick reading anyway, if not, please let me know what is unclear and I can try to make it better.

jcmartins commented:
Please upgrade the README.md

carlpett commented:
@jcmartins I've now added `--collectors.disable-defaults`, was that the part you were missing?

SuperQ commented:
I would prefer to keep the "namespace" the same, so it should be `--collector.disable-defaults` to match the rest of the `--collector.XXX` flags.  What do others think?

carlpett commented:
@SuperQ The rationale was from looking at the other flag which affects "all" collectors, `--collectors.print`, but I've no strong feelings about it.

SuperQ commented:
Ahh, I should have noticed that earlier.  I think with this change, `--collectors.print` is now obsolete since we have a flag per collector.  We should remove it.

carlpett commented:
Hm, I still find it somewhat useful for diagnostic purposes, but sure, I can remove it.

carlpett commented:
@SuperQ @matthiasr @discordianfish Is the removal of `--collectors.print` the only remaining item here? All agree that it should go?

discordianfish commented:
Yes, still need to give it a more thorough review but in general it looks good!

grobie commented on collector/collector.go:
```diff
@@ -15,20 +15,136 @@
+	collectors := make(map[string]Collector)
+	for key, state := range collectorStates {
+		enable := false
+		// Enable the collector if it has been enabled by a flag, OR if it is enabled by default, and the defaults are not disabled
```
Please wrap at 80 chars, end sentence on dot.

SuperQ commented:
@grobie No, that's not a good design pattern that got us here in the first place.  I say we drop the "default off" collector feature and people can just live with reading the docs/`--help` when deploying new releases.

This is the way it works in the `mysqld_exporter` and we've had no problems with it this way.

grobie commented:
@SuperQ Can you please elaborate on your reasoning and provide arguments?

I don't see what part of the proposal is related to the issue of the original implementation of the flags, which required to set the full list of enabled collectors, while the proposed one allows exactly to enable/disable additional collectors one by one. It's also exactly what was requested in the original issue in #216.

> I say we drop the "default off" collector feature and people can just live with reading the docs/--help

:-1: that results in new collectors being enabled silently with new releases without manually adding new disable flags for each of them if someone only wants to run a node_exporter with a specified subset of collectors.

SuperQ commented:
@grobie 

* We want one generally good way to do a thing, in this case enable/disable various collectors.  By doing what you suggest, we now have a very complicated way for users to shoot themselves in the foot.
* We need to be able to "silently" enable collectors by default in order to provide good user experience.  The primary example of this is the `cpu` collector.  We now have separated it from the `stat` collector.  If a user is working with the old flag method, they will now "silently" drop CPU metrics if they don't read the docs/release notes and add it to their custom list.

By having a simple set of boolean flags, we provide a clear way to configure the exporter.  What you are suggesting is very non-standard and complicated.

grobie commented:
Ben and I just had a brief chat. What he actually finds most problematic is the coupling of flags by having an option to disable default flags. I actually fully agree with that. We agreed to have a chat about this next week, I'll probably organize something for the three / interested of us @carlpett.

carlpett commented:
@grobie Sounds good! 

>The `--no-....enabled` variant reads quite awkward.
This is a kingpin feature, actually, boolean flags auto-generate a `--no-<flag>` flag. This is also a fairly common pattern in general unix tools, although admittedly it doesn't really mix that well with the namespacing convention we have.

SuperQ commented:
@grobie @carlpett @discordianfish, I'm mostly available this week to have a quick hangouts call to finish up the discussion of this change.  Does Monday, 2017-09-25 @ 16:00 UTC / 18:00 Berlin work?

discordianfish commented:
Yes, should be able to make that. Would like to finish this off soon too.

carlpett commented:
Yes, that should work.

SuperQ commented:
You can either email me, or join the hangouts link at 18:00: https://plus.google.com/hangouts/_/gitlab.com/node-exporter

carlpett commented:
As a recap of the meeting, here are my notes on actions:
- Remove the default-disabling flag
- Make it clear in `--help` which collectors are enabled by default
- Shorten the collector flags to `--collector.<name>` rather than `--collector.<name>.enabled`, which also makes the `--no-...` flag less awkward
- Sort the `--help` output.

The first three are easy, and were finished more or less immediately. However, as discussed briefly on IRC, the sorting is problematic. I tried using the next version of Kingpin, where it is possible, but since `common/log` references the gopkg.in-type, we get a different type so `log.AddFlags` doesn't work.

There are a few options:
1. Back-port the necessary changes for sorting into kingpin.v2, in the vendor folder. Not very nice.
2. Upgrade all the prometheus projects to v3-unstable, or a pinned commit from the master branch (so the import path will be the same). It took pretty long time go fix all the importers etc the first time, so probably not realistic
3. Re-implement `log.AddFlags` locally in node_exporter
4. Implement the usage output ourselves

Any preferences among the above? @SuperQ @grobie @discordianfish

discordianfish commented:
I think we need a solution for all promtheus components but I also feel like this is(/can be) out of scope of this PR. While I agree that the help output looks awkward, I still think it would be fine to still merge it like that and fix the grouping/sorting later.

SuperQ commented:
I agree with @discordianfish, we do not want to deviate too far from the common Prometheus components.  We can defer the sorted/nice output until later.

carlpett commented:
Ok, good. I updated with the other things now. 
The wifi and qdisc collectors had test-mode flags that conflicted with this new format, so I renamed them `--collector.{wifi,qdisc}.fixtures`. I also note that there are `--collector.procfs` and `--collector.sysfs` that are a bit confusing now (since they are not collectors), maybe those should be changed?

carlpett commented:
Not sure about the test failure. Timeout?

SuperQ commented:
It looks like this change also needs to be re-based.

carlpett commented:
Rebased now. 
As an aside, I cannot run the end-to-end tests locally since my company-issued laptop is a Windows machine, which does not allow :s in file names (created in the sysfs fixtures). Are those paths crucial to the tests, or could they be dropped?

SuperQ commented:
We should have fixed the windows filename issues with the `ttar` tool.

SuperQ commented:
Oh, sorry, I was confused.  We fixed the code so you can check it out, but tests will still not run on windows because it needs to unpack the `ttar` fixtures which contain windows-incompatible file names.  There is currently no way around this.

discordianfish commented:
> I also note that there are --collector.procfs and --collector.sysfs that are a bit confusing now

Agreed, we should find a better name for these. I'd say either `--collector.global.procfs` or `--global.procfs` or something?

Beside that, LGTM!

carlpett commented:
That naming seems fine to me!
@SuperQ @grobie Any feedback on the procfs/sysfs naming from @discordianfish?

SuperQ commented:
No objection to the naming.

grobie commented:
What about just `--procfs` and `--sysfs`? Global is pretty generic, what does that namespace give us in the end?

SuperQ commented:
@grobie Good idea, I like that one best.

carlpett commented:
If we still want some namespace, how about `path`? So `--path.procfs` for example.
Or we could do without

discordianfish commented:
@carlpett I like that, let's go with `--path.`.

discordianfish commented:
@SuperQ Guess you're fine with that too? IMO better than no namespace. We might have more path options in the future so IMO this makes sense.

grobie commented:
SG.

On Thu, Sep 28, 2017 at 11:38 AM discordianfish <notifications@github.com>
wrote:

> @SuperQ <https://github.com/superq> Guess you're fine with that too? IMO
> better than no namespace. We might have more path options in the future so
> IMO this makes sense.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/node_exporter/pull/640#issuecomment-332784028>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaG0j0ufrKymz2TFJaYw1nIz6TObvks5sm2kxgaJpZM4O1qSn>
> .
>

SuperQ commented:
Yes, `--path.` is approved.

grobie commented on README.md:
```diff
@@ -16,7 +16,8 @@ The [WMI exporter](https://github.com/martinlindhe/wmi_exporter) is recommended
 
-Which collectors are used is controlled by the `--collectors.enabled` flag.
+Collectors are enabled by providing a `--collector.<name>` flag.
+Collectors that are enabled by default can be disabled by providing a `--no-collector.name` flag.
```
For consistency, this should be `<name>`

grobie commented on collector/collector.go:
```diff
@@ -15,20 +15,120 @@
+type NodeCollector struct {
+	Collectors map[string]Collector
+}
+
```
NewNodeCollector should have a docstring as it's exported.

carlpett replied:
Hm, perhaps it shouldn't be exported? Not much point to it?

carlpett replied:
Eh, I meant that the `NodeCollector` struct possibly shouldn't be exported. I'll add a doc string to `NewNodeCollector`.

grobie commented on node_exporter.go:
```diff
@@ -30,103 +27,15 @@ import (
-		printCollectors   = kingpin.Flag("collectors.print", "If true, print available collectors and exit.").Bool()
+		listenAddress   = kingpin.Flag("web.listen-address", "Address on which to expose metrics and web interface.").Default(":9100").String()
+		metricsPath     = kingpin.Flag("web.telemetry-path", "Path under which to expose metrics.").Default("/metrics").String()
+		printCollectors = kingpin.Flag("collectors.print", "If true, print available collectors and exit.").Bool()
```
Do we still need this? All available collectors are already visible when showing --help.

discordianfish replied:
Agree, that should be dropped.

SuperQ replied:
Yes, I thought we had talked about dropping this previously.

discordianfish commented:
LGTM beside @grobie's comments.

discordianfish commented:
Awesome, thanks @carlpett for the contribution!
